java/flight/src/main/java/org/apache/arrow/flight/FlightClient.java

```diff
@@ -235,54 +253,74 @@ public void onCompleted() {
 
         };
 
-    asyncServerStreamingCall(call, ticket.toProtocol(), clientResponseObserver);
+    ClientCalls.asyncServerStreamingCall(call, ticket.toProtocol(), clientResponseObserver);
     return stream;
   }
 
-  private static class SetStreamObserver<T> implements StreamObserver<T> {
-    private final SettableFuture<T> result = SettableFuture.create();
-    private volatile T resultLocal;
+  private static class SetStreamObserver implements StreamObserver<Flight.PutResult> {
+
+    private final CompletableFuture<Void> result;
+    private final BufferAllocator allocator;
+    private final StreamListener<PutResult> listener;
```

Comment:
My ask was slightly different. I was proposing that it doesn't make sense to have a completeable future and a listener in the same class. I'm struggling with the pattern where you're doing a double call in the on* event handling calls.

Reply:
Ah, sorry - to be clear, you'd like to essentially merge `PutObserver` and `SetStreamListener`? That sounds reasonable.

Reply:
It's weird to have listener > future > listener. Why not just stack the two listeners and the second one can do the future result (if it wants to go from listener > future)?

Reply:
I guess the issue stems from trying to expose the result in two places; asynchronously via an optional listener given when starting DoPut, and through a future accessible on the listener used to write data during a DoPut. Maybe instead of the optional user-supplied listener defaulting to a no-op one, we can provide an easy way to supply one that converts the result to a future. That is, the current API looks like

```java
final FlightClient.ClientStreamListener writer = client.startPut(descriptor, root, new StreamListener<>() {
    // ....
    @Override
    public void onCompleted() {
    }
});
// call writer.putNext...
writer.getResult(); // implicitly, we have to complete this future, while providing results to the listener above
```
and there is duplication in receiving the events from the server, instead we could have

```java
final FutureListener listener = new FutureListener(); // {
    // optionally override onNext to get application metadata
// };
final FlightClient.ClientStreamListener writer = client.startPut(descriptor, root, listener);
// call writer.putNext()...
listener.getResult();
```

Is that cleaner? There's a bit more boilerplate, but we could get rid of that if we _require_ the listener to be a FutureListener, create a default instance if the client-supplied listener is null, and have `writer.getResult()` just pass through to `listener.getResult`()`.